### PR TITLE
[TASK] Fix support for thumbnail with fixed width and height

### PR DESCRIPTION
**Pull request by dfeyer**
Neos 2.0 introduce a regression that break thumbnails with fixed height
and width (if the ratio of the original image is different than the
ratio of the Thumbnail).

This change allow to set fixed width and height in Image ViewHelpers, TS
Objects and change the internal API. This change is currently breaking
(internal API only) but BC if you use only the ViewHelpers or the TS objects.

This change introduce ```ThumbnailConfiguration``` value object to make
the method signature more readable and make the breaking change more visible.

This change change the database structure for thumbnail entity to store
a JSON representation of the configuration and a hash of this configuration.
The hash is used to check if a thumbnail exist with the same configuration.
By default NULL value are not stored in the database to allow change in the
ThumbnailConfiguration without invalidating all the generated thumbnails.

This change require a database migration:

```
flow doctrine:migrationexecute --version 20150913173832 --direction up
```

This change also simulate to behaviour of Neos 1.2 for BC, where
this snippet will generate a square image of 200x200:

```
<typo3.media:image image="{image}" maximumWidth="200"
maximumHeight="200" allowCropping="true" alt="{title}" />
```

In Neos 2.0, this snipped should be adjusted to:

```
<typo3.media:image image="{image}" width="200" height="200" allowCropping="true"
alt="{title}" />
```

You can unset the settings ```TYPO3.Media.behaviourFlag``` to disable the
Neos 1.2 behaviour compatibility. Previously Neos can generate fixed
dimensions by setting maximumWidth and maximumHeight. The new behaviour
introduced by Neos 2.0 is to use width and height.

Resolves: NEOS-1511

### CODE REVIEW

File: TYPO3.Media/Classes/TYPO3/Media/ViewHelpers/ImageViewHelper.php [hunk 1]
```diff
@@ -103,24 +103,27 @@ public function initializeArguments() {
...
 	 *
 	 * @return string an <img...> html tag
 	 */
-	public function render(ImageInterface $image = NULL, $maximumWidth = NULL, $maximumHeight = NULL, $allowCropping = FALSE, $allowUpScaling = FALSE) {
+	public function render(ImageInterface $image = NULL, $width = NULL, $maximumWidth = NULL, $height = NULL, $maximumHeight = NULL, $allowUpScaling = FALSE) {
 		if ($image === NULL && $this->hasArgument('asset')) {
 			$image = $this->arguments['asset'];
 		}
-		$thumbnailData = $this->assetService->getThumbnailUriAndSizeForAsset($image, $maximumWidth, $maximumHeight, $allowCropping, $allowUpScaling);
+		$allowCropping = $this->hasArgument('allowCropping') && $this->arguments['allowCropping'] !== NULL ? $this->arguments['allowCropping'] : is_integer($width) && is_integer($height);
```
Comment (dfeyer) [inline TYPO3.Media/Classes/TYPO3/Media/ViewHelpers/ImageViewHelper.php]: Not sure if it's the correct way to continue support for allowCropping

File: TYPO3.Neos/Resources/Private/TypoScript/Prototypes/ImageUri.ts2 [hunk 2]
```diff
@@ -2,7 +2,9 @@
 prototype(TYPO3.Neos:ImageUri) {
 	@class = 'TYPO3\\Neos\\TypoScript\\ImageUriImplementation'
 	maximumWidth = 2560
```
Comment (dfeyer) [inline TYPO3.Neos/Resources/Private/TypoScript/Prototypes/ImageUri.ts2]: I think this magic number should be set to NULL in this case the image w/h will be used automatically

File: TYPO3.Neos/Resources/Private/TypoScript/Prototypes/ImageUri.ts2 [hunk 3]
```diff
@@ -2,7 +2,9 @@
...
+	width = NULL
 	maximumHeight = 2560
```
Comment (dfeyer) [inline TYPO3.Neos/Resources/Private/TypoScript/Prototypes/ImageUri.ts2]: I think this magic number should be set to NULL in this case the image w/h will be used automatically

File: TYPO3.Neos/Resources/Private/TypoScript/Prototypes/ImageUri.ts2 [hunk 4]
```diff
@@ -12,11 +14,15 @@ prototype(TYPO3.Neos:ImageUri) {
 prototype(TYPO3.Neos:ImageTag) < prototype(TYPO3.TypoScript:Tag) {
 	asset = 'pass-the-media-asset'
 	maximumWidth = 2560
```
Comment (dfeyer) [inline TYPO3.Neos/Resources/Private/TypoScript/Prototypes/ImageUri.ts2]: I think this magic number should be set to NULL in this case the image w/h will be used automatically

File: TYPO3.Neos/Resources/Private/TypoScript/Prototypes/ImageUri.ts2 [hunk 5]
```diff
@@ -12,11 +14,15 @@ prototype(TYPO3.Neos:ImageUri) {
...
+	width = NULL
 	maximumHeight = 2560
```
Comment (dfeyer) [inline TYPO3.Neos/Resources/Private/TypoScript/Prototypes/ImageUri.ts2]: I think this magic number should be set to NULL in this case the image w/h will be used automatically

File: TYPO3.Neos.NodeTypes/Resources/Private/TypoScript/Root.ts2 [hunk 6]
```diff
@@ -41,9 +41,11 @@ prototype(TYPO3.Neos.NodeTypes:Menu) {
 # Image TS Object
 prototype(TYPO3.Neos.NodeTypes:Image) {
 	maximumWidth = 2560
```
Comment (dfeyer) [inline TYPO3.Neos.NodeTypes/Resources/Private/TypoScript/Root.ts2]: I think this magic number should be set to NULL in this case the image w/h will be used automatically

File: TYPO3.Neos.NodeTypes/Resources/Private/TypoScript/Root.ts2 [hunk 7]
```diff
@@ -41,9 +41,11 @@ prototype(TYPO3.Neos.NodeTypes:Menu) {
...
+	width = NULL
 	maximumHeight = 2560
```
Comment (dfeyer) [inline TYPO3.Neos.NodeTypes/Resources/Private/TypoScript/Root.ts2]: I think this magic number should be set to NULL in this case the image w/h will be used automatically

File: TYPO3.Media/Classes/TYPO3/Media/Domain/Model/ThumbailConfiguration.php [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+	}
+
+	/**
+	 * @return boolean
+	 */
+	public function isCroppingAllowed() {
+		if ($this->width !== NULL && $this->height !== NULL) {
+			return TRUE;
+		}
+		return $this->allowCropping;
```
Comment (dfeyer) [inline TYPO3.Media/Classes/TYPO3/Media/Domain/Model/ThumbailConfiguration.php]: Cropping is always allowed if width and height is set to respect width / height

File: TYPO3.Media/Classes/TYPO3/Media/Domain/Model/ThumbailConfiguration.php [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+	}
+
+	/**
+	 * @return boolean
+	 */
+	public function isUpScalingAllowed() {
+		if ($this->width !== NULL && $this->height !== NULL) {
+			return TRUE;
+		}
+		return $this->allowUpScaling;
```
Comment (dfeyer) [inline TYPO3.Media/Classes/TYPO3/Media/Domain/Model/ThumbailConfiguration.php]: UpScaling is always allowed if width and height is set to respect width / height

**Comment by dfeyer** [comment]
To test this bugfix, check this gist, based on the Demo site:
https://gist.github.com/dfeyer/b7447367c6463450b411

**Comment by dfeyer** [comment]
:+1: The last version is BC and "simulate" the behaviour of Neos 1.2 to generate cropped thumbnail with fixed width/height (maxW + maxH + allowCropping)

**Comment by dfeyer** [comment]
Just updated the Gist with the full TS + Fluid rendering "test suite", this PR is :+1: for me

For the test suite, I reset the default value for maximumHeight/maximumWidth because it can have weird side effect, because of the fallback to 1.2 behaviour. In Fluid those variables are empty so no problem, but in TS if you set maximumHeight or maximumWidth the cropping will be enabled by default, see ```ThumbailConfiguration.php```. We have different option here

- remove the magic 2560 pixel from the TS
- disable the fallback to 1.2 behaviour (this change will be really breaking in this case)
- live with this (and keep this in your mind to help user) ;)

Please review it carefully.

**Comment by dfeyer** [comment]
Don't merge it ... the value for height / width stored in the database are the real height/width of the thumbnail, not the one used by the ResizeAdjustement ... to this PR create a new thumbnail on each request if configured height/width does not match the real height/width of the generated thumbnail ... holy shit ;)

**Comment by dfeyer** [comment]
:+1: 

Now everything work as expected but this change require a database migration, check the updated PR description or the commit message.

File: TYPO3.Media/Classes/TYPO3/Media/Domain/Model/Thumbnail.php [hunk 10]
```diff
@@ -121,6 +110,22 @@ public function getResource() {
 	}
 
 	/**
+	 * @param ThumbailConfiguration $configuration
```
Comment (aertmann) [inline TYPO3.Media/Classes/TYPO3/Media/Domain/Model/Thumbnail.php]: @return void

File: TYPO3.Media/Migrations/Mysql/Version20150913173832.php [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+			$allowUpScaling = isset($configuration['allowUpScaling']) ? $configuration['allowUpScaling'] : NULL;
+			$allowUpScaling = $allowUpScaling ? 1 : 0;
+			$types = [\PDO::PARAM_NULL];
+			$this->addSql("UPDATE typo3_media_domain_model_thumbnail SET maximumwidth = ?, maximumheight = ?, ratiomode = ?, allowupscaling = ?  WHERE persistence_object_identifier = ?", [$maximumWidth, $maximumHeight, $ratioMode, $allowUpScaling, $thumbnailInfo['persistence_object_identifier']], $types);
+		}
+
+		$this->addSql("ALTER TABLE typo3_media_domain_model_thumbnail DROP INDEX originalasset_configurationhash");
+		$this->addSql("ALTER TABLE typo3_media_domain_model_thumbnail DROP configuration, DROP configurationhash");
+	}
+}
```
Comment (aertmann) [inline TYPO3.Media/Migrations/Mysql/Version20150913173832.php]: a migration for Postgres is needed as well

File: TYPO3.Neos.NodeTypes/Resources/Private/TypoScript/Root.ts2 [hunk 12]
```diff
@@ -41,9 +41,11 @@ prototype(TYPO3.Neos.NodeTypes:Menu) {
...
+	height = NULL
 	imageClassName = ${q(node).property('alignment') ? ('typo3-neos-alignment-' + q(node).property('alignment')) : ''}
-	allowCropping = FALSE
+	allowCropping = NULL
```
Comment (kitsunet) [inline TYPO3.Neos.NodeTypes/Resources/Private/TypoScript/Root.ts2]: Why this change? Everywhere else we boolean default it

**Comment by dfeyer** [comment]
PSR2 rebase, now work on the ImageTest package

**Comment by hlubek** [comment]
@dfeyer Have a look at https://github.com/sandstorm/macosx-with-ansible regarding the setup. We use it for all devs in our agency.

**Comment by dfeyer** [comment]
Last change introduce a new setting to disable the Neos 1.2 compatibility, and add a log message to show that the current installation used a deprecated configuration.

We just miss the PostgreSQL migration and we are done

**Comment by dfeyer** [comment]
:+1: merge ready for me, just added the postgres migration by @kitsunet

**Comment by kitsunet** [comment]
:+1: If no one objects I would merge this in 24h

**Comment by hlubek** [comment]
:+1: Fixed some text and added a check to log deprecation messages only once

**Comment by hlubek** [comment]
Merged with 76caa1cd9e917a76dc39f4ea9e66c1f64e437558

File: TYPO3.Neos.NodeTypes/Resources/Private/Templates/NodeTypes/Partials/Image.html [hunk 13]
```diff
@@ -5,12 +5,10 @@
 		<f:then>
 			<f:if condition="{link}">
 				<f:then>
-					<a href="{link}"><media:image asset="{image}" alt="{alternativeText}" title="{title}" maximumWidth="{maximumWidth}" maximumHeight="{maximumHeight}" allowCropping="{allowCropping
-}" allowUpScaling="{allowUpScaling}" /></a>
+					<a href="{link}"><media:image asset="{image}" alt="{alternativeText}" title="{title}" width="{width}" maximumWidth="{maximumWidth}" height="{height}" maximumHeight="{maximumHeight}" allowUpScaling="{allowUpScaling}" /></a>
```
Comment (dimaip) [inline TYPO3.Neos.NodeTypes/Resources/Private/Templates/NodeTypes/Partials/Image.html]: @dfeyer, but why did you remove `allowCropping` from here? Cropping doesn't work without it, obviously.
